integration-tests/integration_test.go

```diff
@@ -0,0 +1,1021 @@
+	}
+
+	issuerResp, err := client.Do(issuerReq)
+	if err != nil {
+		t.Logf("WARNING: Failed to retrieve issuer info: %v", err)
+	} else {
+		defer issuerResp.Body.Close()
+
+		var issuerInfo struct {
+			PublicKey string `json:"public_key"`
+		}
+
+		err = json.NewDecoder(issuerResp.Body).Decode(&issuerInfo)
+		if err != nil {
+			t.Logf("WARNING: Failed to decode issuer response: %v", err)
+		} else {
+			t.Logf("Current active key: %s", issuerInfo.PublicKey)
+		}
+	}
+
+	t.Log("Attempting to redeem each token...")
+	succeeded := 0
+	for i, ti := range tokenInfos {
+		t.Logf("Processing token %d/%d (signed with key: %s)...",
+			i+1, len(tokenInfos), ti.SignedKey)
+
+		payload := "test"
+		signature, err := ti.UnblindedToken.DeriveVerificationKey().Sign(payload)
+		if err != nil {
+			t.Logf("WARNING: Failed to sign payload for token %d: %v", i+1, err)
+			continue
+		}
+
+		redeemRequest := blindedTokenRedeemRequest{
+			Payload:       payload,
+			TokenPreimage: ti.UnblindedToken.Preimage(),
+			Signature:     signature,
+		}
+
+		jsonData, err := json.Marshal(redeemRequest)
+		if err != nil {
+			t.Logf("WARNING: Failed to marshal redemption request for token %d: %v", i+1, err)
+			continue
+		}
+
+		url := fmt.Sprintf("http://cbp:2416/%s/blindedToken/%s/redemption/", which, issuerName)
+		t.Logf("Sending redemption request to: %s", url)
+
+		req, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
+		if err != nil {
+			t.Logf("WARNING: Failed to create HTTP request for token %d: %v", i+1, err)
+			continue
+		}
+
+		req.Header.Set("Content-Type", "application/json")
+		resp, err := client.Do(req)
+		if err != nil {
+			t.Logf("WARNING: Failed to make HTTP request for token %d: %v", i+1, err)
+			continue
+		}
+
+		body, _ := io.ReadAll(resp.Body)
+		resp.Body.Close()
+
+		if resp.StatusCode == http.StatusOK {
+			t.Logf("SUCCESS: Token #%d successfully redeemed (status: %s)", i+1, resp.Status)
+			succeeded++
+
+			// Test duplicate detection
+			t.Logf("Testing duplicate detection for token #%d...", i+1)
+
+			req2, err := http.NewRequest("POST", url, bytes.NewBuffer(jsonData))
+			if err != nil {
+				t.Logf("WARNING: Failed to create duplicate HTTP request: %v", err)
+				continue
+			}
+
+			req2.Header.Set("Content-Type", "application/json")
+			resp2, err := client.Do(req2)
+			if err != nil {
+				t.Logf("WARNING: Failed to make duplicate HTTP request: %v", err)
+				continue
+			}
+
+			defer resp2.Body.Close()
+
+			if resp2.StatusCode == http.StatusConflict {
+				t.Logf("SUCCESS: Duplicate correctly detected (status: %s)", resp2.Status)
+			} else {
+				t.Logf("ERROR: Duplicate NOT correctly detected (expected 409 Conflict, got %d: %s)",
+					resp2.StatusCode, resp2.Status)
+			}
+
+			require.Equal(t, http.StatusConflict, resp2.StatusCode,
+				"Duplicate redemption should be blocked with 409 Conflict")
+		} else {
+			// Some failures are expected, especially for tokens with future keys
```

Comment:
question: if the tests are not deterministic how do we know things are working?

It might be beneficial to move these calls into the specific tests and try and alleviate these issues.